Plug : Propagate dirtiness for all parent changes

johnhaddon opened the pull request:
When plugs are added and/or removed from a node, we need to call `DependencyNode::affects()` so that we can propagate dirtiness to any affected outputs. To date, we have only done this for plugs with the `Dynamic` flag set, where `Dynamic` means roughly "I have been added outside of the node constructor, so I will need to be serialised separately". The primary reason for this restriction was that without it we would end up calling `DependencyNode::affects()` on nodes which hadn't yet finished construction, and inevitably crash as a result.

In #3807 we introduced an optimisation that is much more sensitive to accurate dirty propagation, and @danieldresser-ie raised the possibility that the `Dynamic` flag might not adequately identify all plugs added outside of a constructor.

>The only real concern I have is with the convention that, if I am understanding correctly, it is now undefined behaviour to add a child plug which isn't Dynamic outside of a constructor. 
>
>I can't think off-hand where we're currently doing this at ImageEngine, but if you've got a Python node that stores it's state in a non-network way, and then updates an internal network to reflect changes made to it with buttons or something, and doesn't serialize this internal network, then I would have said it was reasonable ( or even more correct ) to use non-dynamic plugs in the internal network. 

Essentially, `Dynamic` is underspecified for our use in controlling dirty propagation. We wanted it to mean "added outside a constructor", but in reality it means "please serialise a constructor for me", and those two things are subtly different. The `Dynamic` flag has been a constant source of subtle bugs anyway, and I'm actively trying to find ways of killing it.

This PR attempts to address Daniel's concern, by propagating dirtiness any time a plug is added or removed outside of a constructor, regardless of the setting of the `Dynamic` flag. This necessitated tracking the completion of construction, and the only way I could find of doing that in Python was to delve into Boost Python internals beyond my level of competence. Please review carefully!

danieldresser-ie commented:
It certainly makes me happy for us to be able to find a solution of some sort to this, and this seems reasonable.

Your only hesitation seemed to just be not being terribly familiar with the Boost Python guts, and I've got zero acquantaince with that, so I'm not sure how reassuring I can be, but looking through this, nothing looks wrong to me, and it's not as scary looking as I was fearing it might be.

Nothing about this activates my "might weirdly fail non-deterministicly" spider senses ... that metaclass setup stuff, while I don't fully understand it, looks like it would probably crash immediately, so if the tests pass on all platforms, it seems to me like it's probably a usable solution.

The current test failure is a build failure on Mac - maybe GafferVDBModule isn't declaring a dependency on the base lib with this new stuff?  I've noticed before that the Mac CI build requires you to explicitly declare all the libs you're using, whereas we seem to get away with an indirect link on other platforms.

johnhaddon commented:
I've pushed a couple of updates :

- Fixed GafferVDB linking on Mac.
- Added missing GAFFERBINDINGS_API tag on DependencyNodeWrapperBase. This omission was causing the `dynamic_cast` to fail on Mac.
- Added a couple of tests to show that this is all working as intended.

> The Python metaclass stuff looks like the correct solution to me. The implementation seems sane from what I've read about the configuration of g_dependencyNodeMetaclasss in correlation with my last adventures into __metaclass__ some moons ago.

It does worry me a bit that we're committing to a solution that relies on changing the metaclass that Boost uses, which isn't something they explicitly accommodate. But to my naive eyes the code does seem fairly simple and logical, and we do have decent test coverage. Since the whole world seems to be slowly moving to PyBind11, and we might eventually be forced that way by Imath etc I also took a look at that, and they seem to provide explicit support for custom metaclasses, so that's one less worry. And the long term prospects for `makeRefCounted()` and `postConstructor()` do seem to provide broader motivation.

johnhaddon commented:
I mentioned yesterday that I wanted to do a bit of profiling to check that this doesn't affect loading speeds. I can't find any evidence that it does, so I think this is good to go now. The one test failure is our friend the Mac docs build hang, and totally unrelated to this PR. @danieldresser-ie, could you do a quick sanity check on my latest update and merge if it looks OK?
